Disallow thread enable/disable JVMTI_EVENT_SAMPLED_OBJECT_ALLOC

JasonFengJ9 opened the pull request:
### Disallow thread enable/disable JVMTI_EVENT_SAMPLED_OBJECT_ALLOC ###

`JVMTI_EVENT_SAMPLED_OBJECT_ALLOC` is enabled or disabled globally to match RI behaviour.
Throws `JVMTI_ERROR_ILLEGAL_ARGUMENT` if there is an attempt to thread enable/disable.

**Note**: this is to address a `JTReg Test Failure - serviceability/jvmti/HeapMonitor/MyPackage/HeapMonitorEventsForTwoThreadsTest.java`:
```
Exception in thread "main" java.lang.RuntimeException: Sampling event is thread enabled, that is unexpected.
	at MyPackage.HeapMonitorEventsForTwoThreadsTest.main(HeapMonitorEventsForTwoThreadsTest.java:46)
```
With this PR, the test has following output:
```
## Set event notifications error: 103
```
which is same as `RI`.

**Additional note**: this test (and the name HeapMonitorEventsFor**TwoThreads**Test.java) is misleading, the failure messages are identical for first and second thread [1]. It might give an impression that the `SetEventNotificationMode()` failed at second thread but indeed failed at first thread and the second thread serves no purpose from my perspective.

`RI` only allows `JVMTI_EVENT_SAMPLED_OBJECT_ALLOC` to be enabled or disabled globally though `JVMTI` specification doesn't include `Sampled Object Allocation` in the list of events that cannot be controlled at the thread level.

`OpenJ9` matches `RI` behaviours.

[1] https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/2a60f1f273a2c29b411ec9c4d9263d8ff5434529/test/hotspot/jtreg/serviceability/jvmti/HeapMonitor/libHeapMonitorTest.c#L762-L773
[2] https://docs.oracle.com/en/java/javase/12/docs/specs/jvmti.html#SetEventNotificationMode

Reviewer: @DanHeidinga 
FYI: @gacholio 

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

gacholio commented:
Should we update the new JVMTI test to test this case instead of relying on jtreg?

JasonFengJ9 commented:
Modified the test to check the thread enable case, also relax `1` for the disable case to allow a sampling event which might escape before `disable()`.
In addition, reverted code format changes, only fixed un-initialized values.

@gacholio please have another look.

gacholio commented on test/functional/cmdLineTests/jvmtitests/src/com/ibm/jvmti/tests/samplingObjectAllocation/soae001.java:
```diff
@@ -46,16 +46,22 @@ public boolean testDefaultInterval() {
 				} else {
 					reset();
-					bytes = new byte[DEFAULT_SAMPLING_RATE];
+					bytes = new byte[DEFAULT_SAMPLING_RATE * 10];
```
Why times 10?

JasonFengJ9 replied:
Current sampling rate might not be exact as the sampling interval specified as per https://github.com/eclipse/openj9/issues/6024.
Increasing the allocation size in hope to avoid false pass due to insufficient sampling.

gacholio replied:
Would it not make more sense to just not write off the end of the array in that case?

JasonFengJ9 replied:
Discussed with @gacholio , the test for no sampling event fired after `disable()` is not needed considering the nature of non-deterministic event generation.
Removed the test in question, please have another look.

gacholio commented:
jenkins test sanity plinux jdk11

gacholio commented:
Test fails:
```
12:55:54  Output from test:
12:55:54   [OUT] *** Testing [1/1]:	testDefaultInterval
12:55:54   [OUT] Allocated a byte array with size 524288
12:55:54   [OUT] com.ibm.jvmti.tests.samplingObjectAllocation.soae001.enable(thread) failed as expected with: -1
12:55:54   [OUT] *** Test took 10 milliseconds
12:55:54   [OUT] FAILED
12:55:54   [OUT] 
12:55:54   [ERR] exception thrown: 
12:55:54   [ERR] type: 2	Error: 103  JVMTI_ERROR_ILLEGAL_ARGUMENT
12:55:54   [ERR] 		Failed to enable JVMTI_EVENT_SAMPLED_OBJECT_ALLOC event
12:55:54   [ERR] 		Location: com/ibm/jvmti/tests/samplingObjectAllocation/soae001.c -> [Java_com_ibm_jvmti_tests_samplingObjectAllocation_soae001_enable():96]
12:55:54   [ERR] 
12:55:54   [ERR] com.ibm.jvmti.tests.util.AgentHardException: 
12:55:54   [ERR] type: 2	Error: 103  JVMTI_ERROR_ILLEGAL_ARGUMENT
12:55:54   [ERR] 		Failed to enable JVMTI_EVENT_SAMPLED_OBJECT_ALLOC event
12:55:54   [ERR] 		Location: com/ibm/jvmti/tests/samplingObjectAllocation/soae001.c -> [Java_com_ibm_jvmti_tests_samplingObjectAllocation_soae001_enable():96]
12:55:54   [ERR] 
12:55:54   [ERR] 	at com.ibm.jvmti.tests.util.ErrorControl.checkErrors(ErrorControl.java:71)
12:55:54   [ERR] 	at com.ibm.jvmti.tests.util.TestSuite.run(TestSuite.java:87)
12:55:54   [ERR] 	at com.ibm.jvmti.tests.util.TestRunner.main(TestRunner.java:60)
```

JasonFengJ9 commented:
@gacholio the test should be good now, please launch another PR build to verify.

gacholio commented:
jenkins test sanity zlinux jdk11

JasonFengJ9 commented:
Failed again, looking at it.

JasonFengJ9 commented:
The test failed even though the harness printed `OK`:
```
Test result: FAILED
Output from test:
 [OUT] *** Testing [1/1]:	testDefaultInterval
 [OUT] Allocated a byte array with size 524288
 [OUT] com.ibm.jvmti.tests.samplingObjectAllocation.soae001.enable(thread) failed as expected with: -1
 [OUT] *** Test took 6 milliseconds
 [OUT] OK
```
This is due to the 
```
[ERR] com.ibm.jvmti.tests.util.AgentHardException: 
 [ERR] type: 2	Error: 103  JVMTI_ERROR_ILLEGAL_ARGUMENT
 [ERR] 		Failed to enable JVMTI_EVENT_SAMPLED_OBJECT_ALLOC event
 [ERR] 		Location: com/ibm/jvmti/tests/samplingObjectAllocation/soae001.c -> [Java_com_ibm_jvmti_tests_samplingObjectAllocation_soae001_enable():96]
```

Changed `SetEventNotificationMode() error` to `softError`, verified that `make _cmdLineTester_jvmtitests_soae_3` passes.

@gacholio please launch another PR build, thanks.
